openhtf/output/servers/dashboard_server.py

```diff
@@ -0,0 +1,187 @@
+"""HTTP server serving a list of stations found via multicast."""
+
+import argparse
+import collections
+import json
+import logging
+import os
+import socket
+import threading
+import time
+
+from openhtf.core import station_api
+from openhtf.output.web_gui_server import pub_sub
+from openhtf.output.web_gui_server import web_gui_server
+from openhtf.output.web_gui_server import web_launcher
+from openhtf.util import data
+from openhtf.util import multicast
+import sockjs.tornado
+import tornado.web
+
+_LOG = logging.getLogger(__name__)
+
+DASHBOARD_SERVER_TYPE = 'dashboard'
+
+StationInfo = collections.namedtuple(
+    'StationInfo',
+    'cell host port station_id status test_description test_name')
+
+
+def _discover(station_discovery_string, **kwargs):
+  """Yields info about station servers announcing themselves via multicast."""
+  for host, response in multicast.send(station_discovery_string, **kwargs):
+    try:
+      result = json.loads(response)
+    except ValueError:
+      _LOG.warn('Received bad JSON over multicast from %s: %s', host, response)
+    try:
+      yield StationInfo(result['cell'], host, result['port'],
+                        result['station_id'], 'ONLINE',
+                        result.get('test_description'),
+                        result['test_name'])
+    except KeyError:
+      if 'last_activity_time_millis' in result:
+        _LOG.debug('Received old station API response on multicast. Ignoring.')
+      else:
+        _LOG.warn('Received bad multicast response from %s: %s', host, response)
+
+
+class StationListHandler(tornado.web.RequestHandler):
+  """GET endpoint for the list of available stations.
+
+  Sends the same message provided by DashboardPubSub.
+  """
+
+  def get(self):
+    self.write(DashboardPubSub.make_message())
+
+
+class DashboardPubSub(pub_sub.PubSub):
+  """Pub/sub endpoint for the list of available stations."""
+  _lock = threading.Lock()  # Required by pub_sub.PubSub.
+  subscribers = set()  # Required by pub_sub.PubSub.
+  last_message = None
+  station_map = {}
+  station_map_lock = threading.Lock()
+
+  def on_subscribe(self, unused_info):
+    """Called by the base class when a client connects."""
+    if self.last_message is not None:
+      self.send(self.last_message)
+
+  @classmethod
+  def update_stations(cls, station_info_list):
+    """Called by the station discovery loop to update the station map."""
+    with cls.station_map_lock:
+
+      # By default, assume old stations are unreachable.
+      for host_port, station_info in cls.station_map.iteritems():
```

Comment:
items()

Reply:
#727 Thanks to your PR this is now six.iteritems(cls.station_map)